added first draft of profiler docs (#111)
spacemeshos/post-rs: docs/profiler.md

# What is the profiler tool

The major goal of profiler tool is to provide an estimation how fast can be the proof created.
Creating a proof in spacemesh depends on few major factors:
* CPU speed
* Disk speed
* Amount of storage initialized

> The major task of the profiler tool is to estimate how fast can be the proof created for a given CPU speed and disk speed.

Reviewer: Actually, IMO there are 2 tasks:
- estimate how much storage one can safely initialize (to be able to generate POST proof later)
- configure proving (nonces and threads) optimally (to best use resources, so that disk and CPU are in balance)

Author: True will adjust.

Author: Done

> The major task of the profiler tool is to estimate how fast can be the proof created for a given CPU speed and disk speed.

Reviewer 2: ```suggestion
The primary goal of the profiler tool is to estimate how fast a PoST proof can be created for a given CPU and disk speed.
```

The best way to get the profiler tool is to get it from the [releases page](https://github.com/spacemeshos/post-rs/releases).

> It's very important to remember that the proof time itself does not need to be optimized against, we just need to generate valid proof in time which on the mainnet is 12h.

Reviewer: ```suggestion
It's important to understand that the proving time itself does not need to be optimized. It's enough to generate proof in a specified time window, which on the mainnet is 12h.
```

## How to run profiler tool

One of the most important part is to set correct path to data file using `--data-file` flag. It's important to set it so the benchmark will use the proper disk(s). Plaese be warned that the contents of that file MIGHT be overwritten.

> By default profiler uses 1GiB sample data you can change that using `--data-size` flag. One can experiment with different sizes to check for the bottlenecks in the hardware.

Reviewer: ```suggestion
By default, profiler uses 1GiB sample data but it can be changed using `--data-size` flag. One can experiment with different sizes to check for bottlenecks in the hardware.
```

It's also important to give specific `--duration`, that defaults to 10 seconds. The longer the duration the more accurate the results will be.

Proving options. There are two important parameters that then are used by go-spacemesh directly during the proof generation. `--threads` and `--nonces`.
The general rule of thumb is the more threads the faster, but because different processors may have different behaviours when all threads are under load please check different values for `--threads` to find the optimal one.

> `--nonces` The more nonces one does in ONE data pass, the bigger chance to find a proof. The downside is the more `--nonces` the more work CPU will need to do.

Reviewer: It should be documented that nonces must be a multiple of 16.

Author: Done

To simplify the estimation of nonces please use the following formula:
in google sheets format
```
1-(1-(1-BINOM.DIST(36;10^9;26/10^9;TRUE)))^{put nonces here}
```
in wolfram alpha format
```
1-(1-(1-CDF[BinomialDistribution[10^9, 26/10^9],36]))^{put nonces here}
```

That formula will output a probability of finding proof in one data pass. Proving part is written in a way that it will do multiple data passes IF needed.

> Please note that in genreal proof speed should almost linearly scale up with the amount of cores (assuming that they're equally fast), but in the same time it will linearlly scale down with the amount of nonces in groups of 16 (32 nonces should be twice as slow as 16)

Reviewer: spelling/grammar:
```suggestion
Please note that in general proof speed should almost linearly scale up with the number of cores (assuming that they're equally fast), but at the same time it will linearly scale down with the number of nonces in groups of 16 (32 nonces should be twice as slow as 16)
```

That's not entirely true, increasing `nonces` will not slow down if the disk is the bottleneck (as the CPU is not working at 100% yet). 

In the same way, increasing `threads` might not give any improvement on a slow disk.

If `speed_gib_s` is slowing down with more noces, then CPU is most likely the limiting factor. Try to add more threads IF needed

### How to see that disk is my limiting factor?
If `speed_gib_s` is not slowing down with more nonces, then disk is most likely the limiting factor. Try to add more nonces IF needed. That should put more load on CPU and limit the probability of reading the data again from the disk.

### How do I find the sweet spot?
That depends, if you want to generate proof as fast as possible because you cannot have your computer working for 12hours then make it work as hard as possible regardless of anything else.

The general rule of thumb is that you want to match your disk speed with that benchmark. If your disk will be faster then CPU will be working on 100% and disk will NOT be fully utilized, if your CPU is way faster than HDD then CPU may be waiting for the data from the disk.

### Do I need to finish the proof asap?
No, you need to be in time, on the mainnet that time is 12hours. As long as you're on time you're good.

## How to use the values

```
"smeshing-proving-opts": {
  "smeshing-opts-proving-nonces": 144,
  "smeshing-opts-proving-threads": 0
},
```

> Place that in your node config. Please note that the values are just an example and you need to use your own values.

Reviewer: People will not know where to place it in the config. Consider specifying the whole path in JSON:
```
{
  "smeshing": {
    "smeshing-proving-opts": {
      "smeshing-opts-proving-nonces": 144,
      "smeshing-opts-proving-threads": 0
    }
  }
}
```

Author: Not sure honestly, that would assume that people know how to merge json. I gave flat part of a file that they have to add (including semicolon at the end)

Reviewer: OK, but where to put it?

Author: hmm ok you convinced me. I'll add "smashing" also :) but I'll drop the top most {} ok?

Author: @poszu check now, changed my mind a bit and tried to make it more explicit.

# What is the profiler tool

The major goal of profiler tool is to provide an estimation how fast can be the proof created.
Creating a proof in spacemesh depends on few major factors:
* CPU speed
* Disk speed
* Amount of storage initialized

The major task of the profiler tool is to estimate how fast can be the proof created for a given CPU speed and disk speed.

The best way to get the profiler tool is to get it from the [releases page](https://github.com/spacemeshos/post-rs/releases).

> It's very important to remember that the proof time itself does not need to be optimized against, we just need to generate valid proof in time which on the mainnet is 12h.

Reviewer 2: Can you explain what you mean with this? I assume that it is not necessary to optimize proofing speed as long as the proof can be generated in the necessary time window?

Author: Sure, you don't need to complete the proof asap, you want to be as much to the end as possible, and as much as it makes sense in MOST cases that will mean the lowest energy consumption.

## How to run profiler tool

> One of the most important part is to set correct path to data file using `--data-file` flag. It's important to set it so the benchmark will use the proper disk(s). Plaese be warned that the contents of that file MIGHT be overwritten.

Reviewer 2: ```suggestion
For accurate results set the path with `--data-file` to the same disk that will be used by the node, otherwise the results of the benchmark might not reflect the actual performance of the node. Please be warned that the contents of that file MIGHT be overwritten.
```

> By default profiler uses 1GiB sample data you can change that using `--data-size` flag. One can experiment with different sizes to check for the bottlenecks in the hardware.

Reviewer 2: ```suggestion
By default the profiler uses 1GiB sample data. This can be changed with the `--data-size` flag. Different sizes can be used to check for bottlenecks of the hardware.
```

> It's also important to give specific `--duration`, that defaults to 10 seconds. The longer the duration the more accurate the results will be.

Reviewer 2: ```suggestion
The duration of the benchmark can be set with `--duration` and defaults to 10 seconds. A longer duration will yield more accurate results.
```

> Proving options. There are two important parameters that then are used by go-spacemesh directly during the proof generation. `--threads` and `--nonces`.

Reviewer 2: I assume "Proving options." is a heading?

```suggestion
There are two parameters that can influence the proofing speed and can be optimized for go-spacemesh: `--threads` and `--nonces`.
```

> The general rule of thumb is the more threads the faster, but because different processors may have different behaviours when all threads are under load please check different values for `--threads` to find the optimal one.

Reviewer 2: ```suggestion
The general rule of thumb is the more threads the faster, but because different processors may have different behaviours when all threads are under load. Please check different values for `--threads` to find the optimal one.
```

> `--nonces` The more nonces one does in ONE data pass, the bigger chance to find a proof. The downside is the more `--nonces` the more work CPU will need to do.

Reviewer 2: ```suggestion
The more nonces are used with `--nonces` in ONE data pass, the bigger the chance to find a valid proof. The downside of using more `--nonces` is a heavier load on the CPU.
```

To simplify the estimation of nonces please use the following formula:
in google sheets format
```
1-(1-(1-BINOM.DIST(36;10^9;26/10^9;TRUE)))^{put nonces here}
```
in wolfram alpha format
```
1-(1-(1-CDF[BinomialDistribution[10^9, 26/10^9],36]))^{put nonces here}
```

> That formula will output a probability of finding proof in one data pass. Proving part is written in a way that it will do multiple data passes IF needed.

Reviewer 2: ```suggestion
This formula gives the probability of finding a valid proof in one data pass. The node will perform multiple passes if needed.
```

> Please note that in genreal proof speed should almost linearly scale up with the amount of cores (assuming that they're equally fast), but in the same time it will linearlly scale down with the amount of nonces in groups of 16 (32 nonces should be twice as slow as 16)

Reviewer 2: ```suggestion
Please note that proving speed linearly scales with the number of cores used (assuming that they're equally fast), but inversely scales with the number of nonces in groups of 16 (32 nonces should be twice as slow as 16). This effect might not manifest until a high number of nonces is used. The reason is that in most setups the hard disk speed will be the limiting factor if a low number of nonces is used.
```

## How to interpret the results

```
./profiler --data-size 1 --threads=1 --data-file data.bin --nonces=64
{
  "time_s": 12.09140029,
  "speed_gib_s": 0.41351703525481415
}
```

Here we see the following
* 1GiB file `data.bin` was used for the benchmark
* 1 thread was used
* 64 nonces were used

The benchmark took 12.09 seconds to complete and the speed was 0.41GiB/s.

> But from previous formula we know that the probability of finding a proof with 64 nonces is 79.39%. Therefore there is 20% chance that we will need to read the data again. Therefore the actual time can be significanly longer.

Reviewer 2: ```suggestion
From the formula above we know that the probability of finding a proof with 64 nonces is 79.39%. Therefore there is a ~20% chance that at least two passes are necessary (and a ~ 0.20^x chance that more than x passes are necessary).
```

```
./profiler --data-size 1 --threads=1 --data-file data.bin --nonces=128
{
  "time_s": 13.152850458,
  "speed_gib_s": 0.22808744078552953
}
```

We see here that the speed dropped by half. That clearly indicates that CPU is our limiting factor.

With 128 nonces there is 95.75% chance to find a proof in one data pass.

```
./profiler --data-size 1 --threads=10 --data-file data.bin --nonces=128
{
  "time_s": 10.331206291,
  "speed_gib_s": 1.8390882405040923
}
```

> With 10 threads we can see that speed increased by 8 times. But again it's important to remember that we don't necessarily need to optimize for the speed, we need to optimize for the probability of finding a proof in time. There are 12 hours to find and submit proof.

Reviewer 2: ```suggestion
With 10 threads we can see that speed increased by 8 times. It is important to optimize for the probability of finding a proof in time rather than optimizing for the speed. On mainnet nodes have a 12 hour window to find and submit a proof.
```

Based on these outputs you need to decide what is the best configuration for your hardware. Please note that the speed of the proof generation is not the only factor.

## Is that all that is happening during the proof generation?

> Actually no, there is one more item which is that for each group of 16 nonces there is computation required. That exists to prevent a certain kind of exploit where an adversary with cheap processing power can replace some storage with much more computation, we’ve introduced a small amount of additional computation to the PoST proving process making this attack too expensive to carry out.

Reviewer 2: I would shorten the whole sentence to something like:

```
Additionally for every group of 16 nonces there is a proof of work - often referred to as `k2pow` - required. It serves as mitigation against some possible attacks by dishonest smeshers.
```

Author: I took it from the annoucement that was given tot the community, wanted to make it clear that it's not a PoW per see but kinda protection against attack.

Reviewer 2: Then instead of "there is a proof of work" make it "there is an additional computation" 🙂

Author: Changed.

> There are plans to delegate that computation to other servers and pay a small fee for that. You can sometimes find it referred as `k2pow`.

Reviewer 2: I wouldn't add future plans to documentation of existing tools.

Author: I'd keep it honestly.

Reviewer 2: This means we are openly committing to delivering it, we shouldn't make promises that we might not be able to keep.

Author: We already commited in release notes and in announcements ;)

Reviewer 2: but imo then it should be 

```
In the future this computation will be able to be delegated to a server for a small fee. You can sometimes find it referred as `k2pow`.
```

otherwise it reads as if it was hearsay 😅

Author: I removed that part.

> But for now you can assume that for EACH group of 64nonces once per proof process the computation required
> For a low-end CPU with hash rate of 500 in [RandomX benchmark](https://xmrig.com/benchmark) that it should take 2mins30seconds per 1Space Unit (1 Space Unit in mainnet is 64GiB). It scales linearly with the hash rate. Please consult single and multicore results from the benchmark.

Reviewer 2: ```suggestion
For every 64 nonces one `k2pow` computation is required. A low-end CPU with a hash rate of 500 in [RandomX benchmark](https://xmrig.com/benchmark) should require 2mins30seconds per 1 SU (1 Space Unit in mainnet is 64GiB). It scales linearly with the hash rate. Please consult single and multicore results from the benchmark.
```
